core/src/main/java/com/linecorp/armeria/common/stream/FixedStreamMessage.java

```diff
@@ -133,32 +133,56 @@ final void notifySubscriberOfCloseEvent(SubscriptionImpl subscription, CloseEven
-        cancelOrAbort(true);
+        cancelOrAbort(true, null);
     }
 
     @Override
     public final void abort() {
-        final SubscriptionImpl currentSubscription = subscription;
-        if (currentSubscription != null) {
-            cancelOrAbort(false);
-            return;
-        }
+        abort0(AbortedStreamException::get);
+    }
 
-        final SubscriptionImpl newSubscription = new SubscriptionImpl(
-                this, AbortingSubscriber.get(), ImmediateEventExecutor.INSTANCE, false, false);
-        subscriptionUpdater.compareAndSet(this, null, newSubscription);
-        cancelOrAbort(false);
+    @Override
+    public final void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        abort0(() -> cause);
+    }
+
+    @Override
+    public final void abort(Supplier<? extends Throwable> causeSupplier) {
+        requireNonNull(causeSupplier, "causeSupplier");
+        abort0(causeSupplier);
     }
 
-    private void cancelOrAbort(boolean cancel) {
+    private void abort0(Supplier<? extends Throwable> causeSupplier) {
+        cancelOrAbort(false, causeSupplier);
+    }
+
+    private void cancelOrAbort(boolean cancel, @Nullable Supplier<? extends Throwable> causeSupplier) {
         final CloseEvent closeEvent;
-        final Sampler<Class<? extends Throwable>> sampler = Flags.verboseExceptionSampler();
         if (cancel) {
-            closeEvent = sampler.isSampled(CancelledSubscriptionException.class) ?
-                         new CloseEvent(new CancelledSubscriptionException()) : CANCELLED_CLOSE;
+            final Throwable cancelCause = CancelledSubscriptionException.get();
+            setCompletionCause(cancelCause);
+            if (cancelCause == CancelledSubscriptionException.INSTANCE) {
+                closeEvent = CANCELLED_CLOSE;
+            } else {
+                closeEvent = new CloseEvent(cancelCause);
+            }
         } else {
-            closeEvent = sampler.isSampled(AbortedStreamException.class) ?
-                         new CloseEvent(new AbortedStreamException()) : ABORTED_CLOSE;
+            // causeSupplier should not be null if cancel == false
+            final Throwable cause = requireNonNull(causeSupplier.get(),
+                                                   "cause returned by causeSupplier is null");
+            if (subscription == null) {
+                final SubscriptionImpl newSubscription = new SubscriptionImpl(
+                        this, AbortingSubscriber.get(cause), ImmediateEventExecutor.INSTANCE, false, false);
+                subscriptionUpdater.compareAndSet(this, null, newSubscription);
+            }
+
+            setCompletionCause(cause);
+            if (cause == AbortedStreamException.INSTANCE) {
+                closeEvent = ABORTED_CLOSE;
+            } else {
+                closeEvent = new CloseEvent(cause);
+            }
         }
         if (closeEventUpdater.compareAndSet(this, null, closeEvent)) {
```

Comment:
Is it possible to use the return value of `setCompletionCause()` instead of performing an additional CAS?